nested interfaces

### The following program `sample.go` triggers an unexpected result

```go
package main

import (
    "io"
    "os"
)

type sink interface {
    io.Writer
}

func newSink() sink { return os.Stdout }

func main() { 
    s := newSink()
    s.Write([]byte("Hello\n"))
}
```


### Expected result

```console
$ go run ./sample.go
Hello
```


### Got

```console
$ yaegi ./sample.go
/Users/marc/sample.go:16:2: panic
run: reflect: reflect.Value.Call using value obtained using unexported field
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0x14000260200, 0x40)
	/Users/marc/sdk/go1.16.6/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0x140004d3bf0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:504 +0x94
panic(0x10376df20, 0x14000076af0)
	/Users/marc/sdk/go1.16.6/src/runtime/panic.go:965 +0x14c
github.com/traefik/yaegi/interp.runCfg.func1(0x140001dc210, 0x140001d4800, 0x140004d39c0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:185 +0x1e8
panic(0x10376df20, 0x14000076af0)
	/Users/marc/sdk/go1.16.6/src/runtime/panic.go:965 +0x14c
reflect.flag.mustBeExportedSlow(0x2b3)
	/Users/marc/sdk/go1.16.6/src/reflect/value.go:238 +0x114
reflect.flag.mustBeExported(...)
	/Users/marc/sdk/go1.16.6/src/reflect/value.go:229
reflect.Value.Call(0x1038487c0, 0x140000327c0, 0x2b3, 0x1400000e510, 0x1, 0x1, 0x140004d38c8, 0x10308e790, 0x1400000e401)
	/Users/marc/sdk/go1.16.6/src/reflect/value.go:336 +0x3c
github.com/traefik/yaegi/interp.callBin.func1(0x1038487c0, 0x140000327c0, 0x2b3, 0x1400000e510, 0x1, 0x1, 0x103764f60, 0x1400000e4e0, 0x197)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1325 +0x50
github.com/traefik/yaegi/interp.callBin.func10(0x140001dc210, 0x1400018e870)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1494 +0x168
github.com/traefik/yaegi/interp.runCfg(0x140001d4800, 0x140001dc210)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:191 +0x8c
github.com/traefik/yaegi/interp.(*Interpreter).run(0x1400036e000, 0x1400008fe00, 0x14000370000)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:122 +0x244
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x1400036e000, 0x1400058a000, 0xf1, 0x16d0c37f0, 0x16, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:596 +0x5e8
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x1400036e000, 0x16d0c37f0, 0x16, 0x1400018c300, 0xf1, 0x0, 0x0, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:418 +0xc4
main.runFile(0x1400036e000, 0x16d0c37f0, 0x16, 0x0, 0x0, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:149 +0x84
main.run(0x140001c0010, 0x1, 0x1, 0x14000182480, 0x14000073ea8)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:112 +0x830
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x324
```


### Yaegi Version

v0.9.19

### Additional Notes

_No response_

--- a/_test/issue-1128.go
+++ b/_test/issue-1128.go
@@ -7,4 +7,5 @@ func main() {
 	println(len(c))
 }
 
-// Output: 1
+// Output:
+// 1

--- a/_test/issue-1187.go
+++ b/_test/issue-1187.go
@@ -1,0 +1,47 @@
+package main
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+type sink interface {
+	io.Writer
+	io.Closer
+}
+
+func newSink() sink {
+	// return os.Stdout	// Stdout is special in yaegi tests
+	file, err := ioutil.TempFile("", "yaegi-test.*")
+	if err != nil {
+		panic(err)
+	}
+	return file
+}
+
+func main() {
+	s := newSink()
+	n, err := s.Write([]byte("Hello\n"))
+	if err != nil {
+		panic(err)
+	}
+	var writer io.Writer = s
+	m, err := writer.Write([]byte("Hello\n"))
+	if err != nil {
+		panic(err)
+	}
+	var closer io.Closer = s
+	err = closer.Close()
+	if err != nil {
+		panic(err)
+	}
+	err = os.Remove(s.(*os.File).Name())
+	if err != nil {
+		panic(err)
+	}
+	println(m, n)
+}
+
+// Output:
+// 6 6

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1655,9 +1655,12 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 					}
 				} else if m, lind, isPtr, ok := n.typ.lookupBinMethod(n.child[1].ident); ok {
 					n.action = aGetMethod
-					if isPtr && n.typ.fieldSeq(lind).cat != ptrT {
+					switch {
+					case isPtr && n.typ.fieldSeq(lind).cat != ptrT:
 						n.gen = getIndexSeqPtrMethod
-					} else {
+					case isInterfaceSrc(n.typ):
+						n.gen = getMethodByName
+					default:
 						n.gen = getIndexSeqMethod
 					}
 					n.recv = &receiver{node: n.child[0], index: lind}

--- a/interp/interp.go
+++ b/interp/interp.go
@@ -304,7 +304,8 @@ func New(options Options) *Interpreter {
 	// fastChan disables the cancellable version of channel operations in evalWithContext
 	i.opt.fastChan, _ = strconv.ParseBool(os.Getenv("YAEGI_FAST_CHAN"))
 
-	// specialStdio allows to assign directly io.Writer and io.Reader to os.Stdxxx, even if they are not file descriptors.
+	// specialStdio allows to assign directly io.Writer and io.Reader to os.Stdxxx,
+	// even if they are not file descriptors.
 	i.opt.specialStdio, _ = strconv.ParseBool(os.Getenv("YAEGI_SPECIAL_STDIO"))
 
 	return &i

--- a/interp/run.go
+++ b/interp/run.go
@@ -6,7 +6,6 @@ import (
 	"errors"
 	"fmt"
 	"go/constant"
-	"log"
 	"reflect"
 	"regexp"
 	"strings"
@@ -1004,6 +1003,7 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 				vv = v.Elem()
 			}
 		}
+		v = getConcreteValue(v)
 		w := reflect.New(wrap).Elem()
 		w.Field(0).Set(v)
 		for i, m := range methods {
@@ -1016,7 +1016,7 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 				if r := o.MethodByName(names[i]); r.IsValid() {
 					w.Field(i + 1).Set(r)
 				} else {
-					log.Println(n.cfgErrorf("genInterfaceWrapper error, no method %s", names[i]))
+					panic(n.cfgErrorf("method not found: %s", names[i]))
 				}
 				continue
 			}
@@ -1800,16 +1800,23 @@ func getMethodByName(n *node) {
 			}
 			val = v
 		}
+		if met := val.value.MethodByName(name); met.IsValid() {
+			getFrame(f, l).data[i] = met
+			return next
+		}
 		typ := val.node.typ
 		if typ.node == nil && typ.cat == valueT {
 			// happens with a var of empty interface type, that has value of concrete type
 			// from runtime, being asserted to "user-defined" interface.
 			if _, ok := typ.rtype.MethodByName(name); !ok {
-				panic(fmt.Sprintf("method %s not found", name))
+				panic(n.cfgErrorf("method not found: %s", name))
 			}
 			return next
 		}
 		m, li := val.node.typ.lookupMethod(name)
+		if m == nil {
+			panic(n.cfgErrorf("method not found: %s", name))
+		}
 		fr := f.clone(!fork)
 		nod := *m
 		nod.val = &nod

--- a/interp/value.go
+++ b/interp/value.go
@@ -341,6 +341,34 @@ func genValueInterface(n *node) func(*frame) reflect.Value {
 	}
 }
 
+func getConcreteValue(val reflect.Value) reflect.Value {
+	v := val
+	for {
+		vi, ok := v.Interface().(valueInterface)
+		if !ok {
+			break
+		}
+		v = vi.value
+	}
+	if v.NumMethod() > 0 {
+		return v
+	}
+	if v.Type().Kind() != reflect.Struct {
+		return v
+	}
+	// Search a concrete value in fields of an emulated interface.
+	for i := v.NumField() - 1; i >= 0; i-- {
+		vv := v.Field(i)
+		if vv.Type().Kind() == reflect.Interface {
+			vv = vv.Elem()
+		}
+		if vv.IsValid() {
+			return vv
+		}
+	}
+	return v
+}
+
 func zeroInterfaceValue() reflect.Value {
 	n := &node{kind: basicLit, typ: &itype{cat: nilT, untyped: true}}
 	v := reflect.New(interf).Elem()

